### DISCUSSION TITLE
Odd Time Signature Support

### DISCUSSION THREAD

**Discussion by melmakes** [SynthstromAudible/DelugeFirmware#2822] [Ideas]
Hi! I was curious, how hard do we think it would be to implement odd time signatures, like 3/4, 7/4, 5/4, and so on? I'm imagining something similar to what the Squarp Hapax does. When you set it to 3/4, each page of a clip is cut off after the 12th 16th note. 

As someone who listens to a lot of music with odd time signatures, this would be super helpful to me. The current way deluge works has workarounds that allow you to kinda hack in odd time signatures, but when it comes to, for example, playing a clip in a performance and wanting that to line up with everything else, it's a high mental load and doesn't always work the way you'd reasonably want it to.

I'd get my hands dirty, learn how to code for Deluge myself, and try to implement this feature, but alas, grad school is kicking my butt right now and I barely have the spare energy to even play my synths, let alone coding mods for them.

**Comment by lakevillain** [comment]
I'm not a programmer, so I'm just interloping here quickly to voice my support. This feature has been discussed on the official Deluge forums since 2019 and I'm sure the community firmware is the only way it's going to happen. This idea of x-ing out the last columns of a page and having that carry through to every page seems sensible, from a user point of view anyway

Someone else on the forums brought up the idea of being able to blank out columns wherever you want on the page. And I guess this could be helpful say if you have a riff in 7/8 so you could do [] [] [] [] [] [] [] X [] [] [] [] [] [] [] X instead of [] [] [] [] [] [] [] [] [] [] [] [] [] [] X X . 🤷‍♂️ But just having something like this at all would be super helpful.

**Comment by MrS3th** [comment]
As a contemporary jazz performer working with the Deluge as a multitool, I would loooove proper odd time support.

I think that actually supporting all of the common time signatures (2/4, 3/4, 4/4, 6/8, 9/8, 12/8) would go a long way in supporting the rarer ones. I would argue that the only entirely supported time signature currently is 4/4, followed by partial support for 12/8, 6/8 and 2/4. 

4/4 fulfills the 3 following conditions:

1. It's properly played by the metronome and has implicitly its most common unit of time set for tempo (a quarter note). 12/8 is also supported in Triplets View (with a dotted quarter note as unit of time for tempo). 2/4 and 6/8 are easily recordable, although odd numbered bars in musical structures will clash with the metronome. The metronome for 3/4 in Triplets View is currently completely unusable (unless you  3/8 your way around it, while dividing your tempo in 2, which does not set the proper unit of time for tempo, while also being inelegant).

2. It has the proper zoom name scheme for optimal grid readability (1 bar = 1 actual bar and so forth). 3/4 and 6/8 are also supported in Triplets View (although the 2 beats zoom for 3/4 and the 4 beats zoom for 6/8 are somewhat messy, but I do get the current compromise).

3. It always has the first column of the grid be occupied by the first beat of a bar in the time signatures denominator's zoom and up (8th note zoom for 6/8). Currently true for every common time signature with the triplets view workaround but for 9/8 (although the Triplets View workaround for 3/4 makes it impossible to have actual triplets in 3/4 while also having the first column always be a first beat, which is a bit silly).

I like the grid implementation of the Triplets View and agree with everyone here. Once notes are entered for triplets, defining its base value, the grayed out areas double as you zoom, as it should. But there would definitely be a benefit in having a separate way to define time signature, which should then affect the metronome, grid layout and zoom definition of a bar. Selecting a time signature could maybe trigger the most common metronome sequence and grid layout to work with it (and ajusting zoom), while each of them still be independently customisable to cover alternate use case??

For Metronome: 
✓-> high click
+-> mid click
=-> low click
• -> silence

The default metronome for 7/8 (or 7/4) could be:
 ✓======

But as an example, it could also be manually customizable to be those other somewhat common values.
✓===+==
✓==+===
✓••+•••
✓•••+••

The default metronome for 9/8 could be:
✓==+==+==

It could also be:
✓••+••+••

But 9/8 can also be irregular, such as
✓========
✓====+===
✓===+====



For Grid Layout:
0 = enabled column
x = grayed out column

In 7, 0000000x0000000x is a pretty clean cut non-wasteful implementation.

9/8 though is pretty much as wasteful as it gets and also has multiple ways to be subdivided on the grid.

The ternary 000x000x000xxxxx would be the default in 8th note zoom.

But then, for an irregular implementation:
000000000xxxxxxx
Or
00000x0000xxxxxx
Or
0000x00000xxxxxx
Or even maybe
000000000x00000 0000x000000000x

5/4!! Space optimisation or clearer musical structure in a 4 bar unit?
00000000000000x for 3 bars
Or
00000x00000xxxxx for 2 bars


If the user could manually gray out columns in clip view, it would be a good start towards implementing the most workable layouts for odd times I think. And every odd time up to groups of 15 could be implemented in this way.

And then, I suppose that the end game of this would be time signature changes in the context of a single clip, that also modulates the metronome and grid layout. I suppose grid layout could be configured on a per page basis? Or even in those cases the metronome could maybe infer time signature from grid layout? Dunno.

A thought, if a given time signature is the same for all armed clips in song view, the metronome could play that time signature. If not, then, a more neutral click could play in song view?

Lastly, I suppose that for time signatures of 6/4 and higher, the jump between zoom level "2 beats" and "1 bar" could become a bit drastic. 11/4, which would be 5 squares in "2 beats" zoom level would go down to 1 square in a single unzoom. I suppose that the idea behind the "2 beats" zoom level in 4/4 is also 1/2 a complete bar. Maybe adding a zoom level that is actually 1/2 a bar when there are 6 quarter notes or more to a bar? :)

**Reply by MrS3th** [reply]
Also! Never having to use the Triplets View workaround for time signature would mean that it might then be possible to implement tuplets independently from time signature, which would also likely make for a better implementation of that feature (as detailled on another discussion post), although, I suppose that the interaction between grid layout for time signature, and grid layout for tuplets could become messy. But it would have no incidence on the metronome as detailled above, yay! (Or we could make polyrythms possible for the metronome also?? Haha. Two independently programmable metronome tracks??)

**Comment by safrosch** [comment]
Just want to jump on and second all the comments made above! Where other groove boxes are completely unusable in regard to odd time signatures or polymetric looping, the deluge could be the pioneer to boldly loop where no synth has gone before. 

On top of the suggestions above, I’d really love to see some support for audio looping in odd time signatures. (1) I’d love to be able to start a loop with no tracks going with a preset track length or multiple of that length. The midiphy Loopa does an amazing job with this for midi looping where you set the time signature in advance and it’ll set the tempo based on the loop you created in your designated time signature. (2) the only way I’ve found to loop in an odd/polymetric manner is by zooming in on the screen resolution to 64th or 32nd notes which effectively changes the record/loop quantization to whatever the screen resolution is. What I’d love to see is a global setting that allows you to change record/loop quantization to different values. For instance if you changed the loop quantization to a quarter note, then whenever you close the loop it’ll close the very next note allowing for you to loop in all different beat values. 

Thanks to the developers for all the work they do!

**Reply by manysounds** [reply]
Very much need audio looping in odd times, and I mean 6/8 “odd”. Yes, dealing with the grid could be messy programming but “simply” disabling the last 4 columns to give the right amount of beats if there’s a “clip time signature” would be nice enough but most importantly the ability to loop a 7/8 would be amazing for my work and one of the last Deluge wish-list items I have.
